### PR TITLE
PR LiRA

### LINKED ISSUE #9: LiRA
# Feature
## Desired Behavior / Functionality
We want to implement LiRA from "Carlini, Nicholas, et al. "Membership inference attacks from first principles." 2022 IEEE Symposium on Security and Privacy (SP). IEEE, 2022."

## What Needs to Be Done
- Create a class LiRA that implements the attack
- Add it to attack factory

## How Can It Be Tested
- Unit test on toy data
- Comparison to RMIA
- Compare to paper results

### LINKED ISSUE #31: Online attacks for LiRA
# Feature

### LINKED ISSUE #32: set FPR statistics
# Feature
## Desired Behavior / Functionality
For given threshold(s) of FPR, print and/or save the TPR for each attack run

## What Needs to Be Done
Implement logging of the TPR for a few given FPRs eg. 0%, 0.1% and 1%

### LINKED ISSUE #57: Shadow_model_handler, training data not sampling from the correct indicies
# Issue
## Problem Description
In the file shadow_model_handler.py ( in the function create_shadow_models ) , when sampling data to be used for training shadowmodels, it will not sample the intended data. It can sample multiple forbidden indicies to use e.g. target train and test data when they are excluded from " attack_data_index "

## Expected Behavior
When sampling training data from shadow models, it can't use excluded indicies

## What Needs to be Done
Make the shadow_model_handler.create_shadow_model() replace the dataset argument "attack_data" with the list "attack_data_index" and  the dataset "population" and use these to sample the shadow model training data. 

## How Can It Be Tested or Reproduced
Use the "get_in_indices_mask" in shadow_model_handler and check that the offline attacks returns 0 True(s)
This indicates that the offline attack has 0 IN- samples. For any offline attack this should hold.

OR check that the indicies that is beeing sampled for "shadow_data_indices" is strictly a subsample of the allowed indicies ( e.g. not containing any target train and test indicies if the are to be excluded )

### LINKED ISSUE #58: Not excluding train and test indicies properly
# Issue
## Problem Description
In the file attack_data.py, the exclusion of target dataset(s), train and test, not is not working correctly

## Expected Behavior
If both flags, include_train_indicies and include_test_indicies is set to false both datasets should be excluded.

## What Needs to be Done
Redefine how the exclusion is done

## How Can It Be Tested or Reproduced
Print the length of the not_allowed_indices, for when the flags are both true and false

### PR DESCRIPTION
**Pull request by henrikfo**
# Description

Summary of changes

* Implementing offline-LiRA ( online LiRA also but commented out due to bug searches, will make another PR when it's gone they are all fixed)
* Implemeted a function that writes down and saves TPRs for a few predefined FPRs for each attack setup.
* Shadow model training data not sampled from the intended distribution, now it does. 
* Redefined how train and test indicies are excluded, found a unexpected issue where is did not exclude accodingly with flags
* Added a function that creates masks for whether or not audit index was used in shadow model training dataset, which then can be used to gather IN- and OUT- sample logits

## Resolved Issues

- [ ] fixes #9 
- [ ] fixes #31 
- [ ] fixes #32 
- [ ] fixes #57 
- [ ] fixes #58 
- [ ] Added Requirement file ( Not updated in the enviorment file, issue is thus not fixed )

## Added feature(s) 

- [ ] Added #56 

## How Has This Been Tested?
Offline LiRA attack performed

### CODE REVIEW

**Comment by henrikfo** [comment]
Please do check whether or not you also had these issues I listed above, if you have any questions please message me

File: leakpro/attacks/mia_attacks/lira.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            shadow_models_logits = self.shadow_models_logits[i, :]
+            out_mean = np.nanmean(shadow_models_logits[~mask])  # Calculate the mean of shadow model logits
+            
+            # Get the logit from the target model for the current sample
+            target_logit = self.target_logits[i]
+    
+            # Calculate the log probability density function value
+            if not self.fixed_variance:
+                out_std = np.nanstd(shadow_models_logits[~mask])
+                # in_std = np.nanstd(shadow_models_logits[mask])
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/lira.py]: in general, I think we should avoid using np.nanstd and np.nanmean. We dont get to see if there are errors because the nans are not visible to us.

**Review by johanos1 — CHANGES_REQUESTED** [review]

File: leakpro/attacks/utils/shadow_model_handler.py [hunk 2]
```diff
@@ -304,3 +305,51 @@ def identify_models_trained_on_samples(self:Self, shadow_model_indices: list[int
...
+            raise ValueError("Index out of range")
+        meta_data = {}
+        with open(f"{self.storage_path}/{self.metadata_storage_name}_{index}.pkl", "rb") as f:
+            meta_data = pickle.load(f)
+        return meta_data
+        
+    def get_shadow_model_metadata(self:Self, num_models:int) -> list:
+        """Load the the shadow model metadata."""
+        metadata = []
+        for i in range(num_models):
```
[the remark this answers is not in the document]
Comment (henrikfo) [reply leakpro/attacks/utils/shadow_model_handler.py:342]: Yes, but it will load the metadata after the all shadowmodels have been loaded/trained. Maybe I am missing something here but this process will be the same whether or not all shadow models or some is trained or loaded from disc

File: leakpro/attacks/mia_attacks/lira.py (hunk 1 above)
Comment (henrikfo) [reply leakpro/attacks/mia_attacks/lira.py]: Yes, for sure. Will remove them

File: leakpro/attacks/utils/shadow_model_handler.py [hunk 3]
```diff
@@ -304,3 +305,51 @@ def identify_models_trained_on_samples(self:Self, shadow_model_indices: list[int
...
+            self.logger.info(f"Loading metadata {i}")
+            metadata.append(self._load_metadata(i))
+        return metadata
+
+    def get_in_indices_mask(self:Self, num_models:int, dataset:np.ndarray) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/utils/shadow_model_handler.py:347]: Ok, are both still needed? Shall we get rid of one of them?

File: leakpro/attacks/utils/shadow_model_handler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/utils/shadow_model_handler.py:342]: yea, let's keep this for now.

File: leakpro/attacks/utils/shadow_model_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (henrikfo) [reply leakpro/attacks/utils/shadow_model_handler.py:347]: Maybe not for the purpose of extracting IN vs OUT logits, one should suffice. I have implemented a simple numba decorator for get_in_indices_mask(), but there should not be any difference in speed between them with the same optimisations.

**Review by johanos1 — APPROVED** [review]
